### DISCUSSION TITLE
Keyboard Qwerty

### DISCUSSION THREAD

**Discussion by HKpro2090** [Bodmer/TFT_eSPI#1616] [Q&A - Sprites]
I m trying to build a keyboard using TFT_espi, I would like to know whether it is possible to use the Button class inside sprite. I don't see any function call to add a button to sprite and later push the sprite when needed.

**Comment by pipi61** [comment]
try this: [https://forum.arduino.cc/t/adafruit-tft-touchscreen-keypad-for-touchscreen/347024](url)
adapted to tft_espi:
[Portrait_Keyboard.ino.txt](https://github.com/Bodmer/TFT_eSPI/files/8008963/Portrait_Keyboard.ino.txt)

**Reply by HKpro2090** [reply]
I get it how he has done but won't it be better if u could draw the buttons in sprite so we can push sprite whenever we want like google keyboard on android. is there any examples of that? In the example you provided he is drawing the buttons or keys directly on to the screen right?

**Reply by videobuff** [reply]
> try this: [https://forum.arduino.cc/t/adafruit-tft-touchscreen-keypad-for-touchscreen/347024](url)
> adapted to tft_espi:
> [Portrait_Keyboard.ino.txt](https://github.com/Bodmer/TFT_eSPI/files/8008963/Portrait_Keyboard.ino.txt)

Dear Istvan,

The 1st link comes up with an 404 error. The textfile however, worked fine.I use a WT32-SCO1 display & esp32 in one device, so i have to rewrite the part for the touch. (it uses FT6336U) 
Question. Varying the height value the keyboard goes up and down on the screen
, but how can i move it to the right ?

![Schermafbeelding 2022-02-09 om 00 43 33](https://user-images.githubusercontent.com/3132953/153094514-39b27692-ce1c-4f7a-ac82-85c925173b7c.jpg)

**Reply by pipi61** [reply]
Sorry if click the link, then bad, but copy the link text, and open, then work
[[https://forum.arduino.cc/t/adafruit-tft-touchscreen-keypad-for-touchscreen/347024](https://github.com/Bodmer/TFT_eSPI/discussions/url)](url)
 the another link: [https://forum.arduino.cc/index.php?topic=359928.0](url)
Make/define new offset variable, KEYLX and KEYY, and add this variable to function X,Y cordinate...
orig: 
`      drawButton(8 + (23 * (x - 3)) + ShiftRight, Height + (30 * y), 20,25); // this will draw the button on the screen by so many pixels
      tft.setCursor(12 + (23 * (x - 3)) + ShiftRight, Height+5 + (30 * y));
`
new:
`drawButton(KEYLX + (KEYDIST * (x - 3)) + ShiftRight, KEYY + (KEYDIST * y), KEYW, KEYH); // this will draw the button on the screen by so many pixels
      tft.setCursor(KEYLX + KEYTOX + (KEYDIST * (x - 3)) + ShiftRight, KEYY + KEYTOY + (KEYDIST * y));
`
.....

**Reply by pipi61** [reply]
I found now my source :)) I modify another constant to variables too. sorry, for (my english) the comment in hungarian somewhere
[TFT_Keypad.ino.txt](https://github.com/Bodmer/TFT_eSPI/files/8028600/TFT_Keypad.ino.txt)
 :))
comment: exit from  the GetKeyPress function, if return pressed... I modify origin, and add another keyboard table too

**Reply by pipi61** [reply]
sorry the forum change the webpage links... don't click, but select the link text, and copy /paste to browser

**Reply by suntchi2018** [reply]
Hi pipi61 i have a display 240x320 touch resistive ST7789_DRIVER I tried to use your keyboard, but it's not centered and I'm not sure, but the dimensions are maybe too big, can you give me some guidance on how to center it in a 240x320 screen?
Thanks in advance for your help
![20230408_062348](https://user-images.githubusercontent.com/40439798/230703098-a935d6cc-073d-4fa7-aace-c05334dd9d84.jpg)

**Reply by dannybirdi** [reply]
Can you please upload a simple sketch for the Touch adaptation?

**Comment by Bodmer** [comment]
The Button class is designed for direct screen writes.  It uses Button coordinates and these are not always the same as screen position for the touch.

You should find a direct write to the screen works fine.
